### PR TITLE
Add additional roles and calculate a specific language in TenantUserManager

### PR DESCRIPTION
**Pull request by welzfabian**

### CODE REVIEW

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 1]
```diff
@@ -735,6 +747,14 @@ protected String computeTenantname(@Nullable WebContext ctx, String tenantId) {
     @Nonnull
     @Override
     protected String computeLang(WebContext ctx, String userId) {
-        return NLS.getDefaultLanguage();
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: shouldn't we rather add a lang field for UserAccount? and a defaultLanguage for Tenant and use those?

File: src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    /**
+     * Adds additonal roles to the given {@link Set} of roles, based on the given {@link Tenant} and {@link UserAccount}.
+     *
+     * @param user   the {@link UserAccount} for which the roles should be calculated
+     * @param tenant the {@link Tenant} for which the roles should be calculated
+     * @param roles  the roles calculated by the framework
+     * @return the {@link Set} of roles, which will be given to the {@link sirius.web.security.UserInfo UserInfo} as roles.
+     */
+    @Nonnull
+    Set<String> addAdditionalRoles(UserAccount<?, ?> user, Tenant<?> tenant, Set<String> roles);
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java]: add a Consumer<String> and remove the mutable set as parameter

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 3]
```diff
@@ -684,9 +689,16 @@ private boolean isWithinInterval(LocalDateTime dateTime, Integer requiredInterva
         Set<String> transformedRoles = transformRoles(roles);
         if (isSystemTenant && transformedRoles.contains(PERMISSION_MANAGE_SYSTEM)) {
             roles.add(PERMISSION_SYSTEM_TENANT);
-            return transformRoles(roles);
+            return applyAdditionalRolesProvider(user, tenant, transformRoles(roles));
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: applyAdditionalRoles(...., roles::add)

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 4]
```diff
@@ -132,6 +132,12 @@
     @Part
     protected static AuditLog auditLog;
 
+    @Part
+    private static AdditionalRolesProvider additionalRolesProvider;
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: PartCollection

Comment (andyHa) [reply src/main/java/sirius/biz/tenants/TenantUserManager.java]: /Parts

File: src/main/java/sirius/biz/tenants/TenantData.java [hunk 5]
```diff
@@ -218,6 +218,11 @@
     public static final Mapping JOURNAL = Mapping.named("journal");
     private final JournalData journal;
 
+    public static final Mapping LANG = Mapping.named("lang");
+    @NullAllowed
```
Comment (welzfabian) [inline src/main/java/sirius/biz/tenants/TenantData.java]: alternative we could set the `lang` in `@BeforeSave` to the `NLS.getDefaultLanguage()` if empty

File: src/main/java/sirius/biz/tenants/UserAccountData.java [hunk 6]
```diff
@@ -86,6 +86,11 @@
     @AutoImport
     private boolean externalLoginRequired = false;
 
+    public static final Mapping LANG = Mapping.named("lang");
+    @Autoloaded
+    @NullAllowed
```
Comment (welzfabian) [inline src/main/java/sirius/biz/tenants/UserAccountData.java]: alternative we could set the `lang` in `@BeforeSave` to the `NLS.getDefaultLanguage()` if empty

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 7]
```diff
@@ -684,9 +692,16 @@ private boolean isWithinInterval(LocalDateTime dateTime, Integer requiredInterva
         Set<String> transformedRoles = transformRoles(roles);
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: shouldn't we applyAdditionalRoles here - before a transform happens?

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 8]
```diff
@@ -684,9 +692,16 @@ private boolean isWithinInterval(LocalDateTime dateTime, Integer requiredInterva
...
         if (isSystemTenant && transformedRoles.contains(PERMISSION_MANAGE_SYSTEM)) {
             roles.add(PERMISSION_SYSTEM_TENANT);
-            return transformRoles(roles);
+            return applyAdditionalRolesProvider(user, tenant, true, transformRoles(roles));
+        }
+        return applyAdditionalRolesProvider(user, tenant, false, transformedRoles);
+    }
+
+    private Set<String> applyAdditionalRolesProvider(U user, T tenant, boolean isSystemTenant, Set<String> roles) {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: _Providers

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 9]
```diff
@@ -732,9 +747,20 @@ protected String computeTenantname(@Nullable WebContext ctx, String tenantId) {
         }
     }
 
+    @Nonnull
+    public List<String> getAvailableLanguages() {
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: javadoc

File: src/main/java/sirius/biz/tenants/UserAccountData.java [hunk 10]
```diff
@@ -86,6 +86,11 @@
     @AutoImport
     private boolean externalLoginRequired = false;
 
+    public static final Mapping LANG = Mapping.named("lang");
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/UserAccountData.java]: javadoc

File: src/main/resources/biz_de.properties [hunk 11]
```diff
@@ -420,3 +420,33 @@ SchedulerData.userId=Anwender
 SchedulerData.userId.help=Wählen Sie hier aus, als welcher Anwender der Job gestartet werden soll. Achten Sie darauf, dass der Anwender nicht gesperrt ist und über alle benötigten Berechtigungen verfügt.
 JobSchedulerLoop.jobLink=Geplante Ausführung
 JobSchedulerLoop.scheduledExecutionInfo=Der Job wurde durch die geplante Ausführung '${entry}' gestartet...
+Language = Sprache
```
Comment (andyHa) [inline src/main/resources/biz_de.properties]: Model.lang

File: src/main/resources/default/templates/biz/tenants/tenant-details.html.pasta [hunk 12]
```diff
@@ -9,6 +10,15 @@
             <w:textfield span="12" name="tenantData_accountNumber" value="@tenant.getTenantData().getAccountNumber()"
                          labelKey="Tenant.accountNumber" required="true"/>
         </div>
+        <div class="row">
+            <w:singleSelect name="tenantData_lang" labelKey="Language">
```
Comment (andyHa) [inline src/main/resources/default/templates/biz/tenants/tenant-details.html.pasta]: labelKey="Model.lang"

File: src/main/resources/default/templates/biz/tenants/user-account-details.html.pasta [hunk 13]
```diff
@@ -11,6 +12,16 @@
                          helpKey="UserAccount.username.help"/>
         </div>
 
+        <div class="row">
+            <w:singleSelect name="userAccountData_lang" labelKey="Language">
```
Comment (andyHa) [inline src/main/resources/default/templates/biz/tenants/user-account-details.html.pasta]: s.a.

File: src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    /**
+     * Adds additonal roles to the given {@link Set} of roles, based on the given {@link Tenant} and {@link UserAccount}.
+     *
+     * @param user           the {@link UserAccount} for which the roles should be calculated
+     * @param tenant         the {@link Tenant} for which the roles should be calculated
+     * @param isSystemTenant indicate wheter the tenant is the system tenant
+     * @param roleConsumer   the consumer for the additonal roles
+     */
+    void addAdditionalRoles(UserAccount<?, ?> user,
+                            Tenant<?> tenant,
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java]: isn't that rendundant to user.getTenant().getValue() ?

File: src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+                            boolean isSystemTenant,
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/AdditionalRolesProvider.java]: is that really necessary - shouldn't we rather supply the current roles set as unmodifyable set for such check - if at all?

File: src/main/java/sirius/biz/tenants/TenantController.java [hunk 16]
```diff
@@ -246,7 +247,8 @@ public void selectTenants(WebContext ctx) {
                             QueryField.contains(Tenant.TENANT_DATA.inner(TenantData.ADDRESS).inner(AddressData.STREET)),
                             QueryField.contains(Tenant.TENANT_DATA.inner(TenantData.ADDRESS).inner(AddressData.CITY)));
 
-        ctx.respondWith().template("templates/biz/tenants/select-tenant.html.pasta", ph.asPage(), isCurrentlySpying(ctx));
+        ctx.respondWith()
+           .template("templates/biz/tenants/select-tenant.html.pasta", ph.asPage(), isCurrentlySpying(ctx));
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantController.java]: code format?

File: src/main/java/sirius/biz/tenants/TenantUserManager.java [hunk 17]
```diff
@@ -144,6 +150,9 @@ protected TenantUserManager(ScopeInfo scope, Extension config) {
         super(scope, config);
         this.systemTenant = config.get("system-tenant").asString();
         this.acceptApiTokens = config.get("accept-api-tokens").asBoolean(true);
+        this.availableLanguages = config.getStringList("availableLanguages").isEmpty() ?
```
Comment (andyHa) [inline src/main/java/sirius/biz/tenants/TenantUserManager.java]: use snake case as all other keys above do

**Pull request merged**